src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -97,169 +135,335 @@ func newCommitLogSource(opts Options, inspection fs.Inspection) bootstrap.Source
+			err := s.bootstrapShardSnapshots(
+				ns.Metadata, accumulator, shard, tr, blockSize,
+				snapshotFilesByShard[shard],
+				mostRecentCompleteSnapshotByBlockShard)
+			if err != nil {
+				return bootstrap.NamespaceResults{}, err
+			}
+		}
 	}
 
+	s.log.Info("read snapshots done",
+		zap.Stringer("took", s.nowFn().Sub(startSnapshotsRead)))
+
+	// Setup the series accumulator pipeline.
 	var (
-		bOpts     = s.opts.ResultOptions()
-		blOpts    = bOpts.DatabaseBlockOptions()
-		blockSize = ns.Options().RetentionOptions().BlockSize()
+		// TODO(r): rename EncodingConcurrency to AccumulateConcurrency.
+		numWorkers = s.opts.EncodingConcurrency()
+		workers    = make([]*accumulateWorker, 0, numWorkers)
 	)
-
-	readCommitLogPred, mostRecentCompleteSnapshotByBlockShard, err := s.newReadCommitlogPredAndMostRecentSnapshotByBlockShard(
-		ns, shardsTimeRanges, snapshotFilesByShard)
-	if err != nil {
-		return nil, err
+	for i := 0; i < numWorkers; i++ {
+		worker := &accumulateWorker{
+			inputCh: make(chan accumulateArg, workerChannelSize),
+		}
+		workers = append(workers, worker)
+	}
+	closedWorkerChannels := false
+	closeWorkerChannels := func() {
+		if closedWorkerChannels {
+			return
+		}
+		closedWorkerChannels = true
+		for _, worker := range workers {
+			close(worker.inputCh)
+		}
 	}
+	// NB(r): Ensure that channels always get closed.
+	defer closeWorkerChannels()
 
 	// Setup the commit log iterator.
 	var (
-		nsID              = ns.ID()
-		seriesSkipped     int
-		datapointsSkipped int
-		datapointsRead    int
-
-		// TODO(rartoul): When we implement caching data across namespaces, this will need
-		// to be commitlog.ReadAllSeriesPredicate() if CacheSeriesMetadata() is enabled
-		// because we'll need to read data for all namespaces, not just the one we're currently
-		// bootstrapping.
-		readSeriesPredicate = func(id ident.ID, namespace ident.ID) bool {
-			shouldReadSeries := nsID.Equal(namespace)
-			if !shouldReadSeries {
-				seriesSkipped++
-			}
-			return shouldReadSeries
-		}
-
 		iterOpts = commitlog.IteratorOpts{
-			CommitLogOptions:      s.opts.CommitLogOptions(),
-			FileFilterPredicate:   readCommitLogPred,
-			SeriesFilterPredicate: readSeriesPredicate,
+			CommitLogOptions:    s.opts.CommitLogOptions(),
+			FileFilterPredicate: s.readCommitLogFilePredicate,
+			// NB(r): ReturnMetadataAsRef used to all series metadata as
+			// references instead of pulling from pool and allocating,
+			// which means need to not hold onto any references returned
+			// from a call to the commit log read log entry call.
+			ReturnMetadataAsRef: true,
 		}
+		startCommitLogsRead = s.nowFn()
 	)
-
+	s.log.Info("read commit logs start")
 	defer func() {
-		s.log.Info("ReadData finished",
-			zap.Int("seriesSkipped", seriesSkipped),
+		datapointsSkipped := 0
+		datapointsRead := 0
+		for _, worker := range workers {
+			datapointsSkipped += worker.datapointsSkipped
+			datapointsRead += worker.datapointsRead
```

Comment:
Could it be useful to track max, min, avg here? Just to get an idea if some encoders end up running really hot for whatever reason

Reply:
Opting to do this later.